rustls/src/suites.rs

```diff
@@ -189,7 +189,10 @@ pub(crate) fn choose_ciphersuite_preferring_server(
 }
 
 /// Return a list of the ciphersuites in `all` with the suites
-/// incompatible with the Groups extension removed.
+/// incompatible with the groups extension `groups_ext` removed.
+///
+/// Leaves `all` unchanged if there is no overlap between `groups_ext`
```

Comment:
This seems like an odd choice. Why not clear it proactively?

Reply:
At the minute we do things in this order: 1) choose a cipher suite, 2) choose a key exchange group. The logic for (1) is the same for both TLS1.2 and 1.3 (happens in `hs.rs`), whereas (2) is version-specific (in `tls12.rs` or `tls13.rs`).

Prior to the FFDHE PR, and because we only ever supported ECDHE suites and key exchange, there was no need to filter cipher suites by key exchange type (recall that for TLS1.2 only, suites are specific to ECDHE or FFDHE, so if there is no prospect of ECDHE being used, those must be removed from the choice of cipher suite.)

So this change is restoring the old behaviour, allowing the version-specific key exchange choice code to do the `NoKxGroupsInCommon` error handling as before.

The alternatives would be:
- to hoist the key exchange choice up into `hs.rs` and fuse it with the suite selection, then pass the selected group into the version-specific code
- keep things as they are, and have some statically unreachable error handling code

Reply:
> to hoist the key exchange choice up into hs.rs and fuse it with the suite selection, then pass the selected group into the version-specific code

Going to see how this one goes.

Reply:
OK! I think this is not too bad. Please re-review.